rfd/0193-access-control-decision-api.md

```diff
@@ -0,0 +1,263 @@
+---
+authors: Alan Parra (alan.parra@goteleport.com), Forrest Marshall (forrest@goteleport.com)
+state: draft
+---
+
+# RFD 0193 - Access Control Decision API
+
+## Required approvers
+
+- Engineering: @rosstimothy
+
+## What
+
+This proposal describes a means of reworking teleport access-control decisions to remove the need to perform
+primary access-control decisions on peripheral agents and to bring teleport's internals more in line with common
+access-control best practices.
+
+## Why
+
+Access-control decisions in teleport are typically made at the agent, and often times a single operation
+will incur many small access-control decisions over the course of its execution. This distributed and innervated
+model of access-control decision making has some benefits. For example, early versions of teleport made it a
+priority to have agents continue to be able to make access-control decisions even when the auth service was
+unavailable for long periods of time. However, there are also serious drawbacks to the distributed and innervated
+model:
+
+- Scalability: Access-control decisions at the agent require that agents have access to all necessary information
+in order to be able to make correct access-control decisions. This forces us to encode a large amount of otherwise
+unnecessary information into certificates, and forces agents to cache all resources relevant to access-control
+decision making (namely roles). In some larger clusters, role replication alone can account for a significant portion
+of all load incurred after an interruption of auth connectivity.
+
+- Compatibility: Because teleport access-control includes the ability to specify unscoped deny rules, there is no
+sane way to represent many new access-control features to older agents s.t. they can continue to make sensible
+decisions. We've typically had to resort to cumbersome and imperfect hacks that negatively impact user experience,
+such as injecting wildcard deny rules into the caches of outdated agents to force them to deny any access that might
+require knowledge of a new feature.
+
+- Maintainability: From both the developer and user perspective, distributed and innervated access-control decision
```

Comment:
nit: you prompted me to look up "innervated" after the third time reading it, to be honest I don't think "and innervated" is adding much here

Reply:
I find the distinction compelling and relevant because teleport access-control decisions aren't *just* distributed.  We also pass around roles/etc and make lots of little decisions at different points/steps within the enforcement logic.  The two phases of work in this RFD distinctly each target one of the two problems.  The refactor phase solves the innervation problem, and the relocate phase solves the distributed problem.

I realize that innervate is not a super commonly used word.  My using it here is definitely letting my background in the biological sciences peek through.  I don't know of a better word for what I'm trying to describe though (the distinction between decision making that happens on many machines, and decision-making that reaches into and touches many places within the enforcement logic being run on a given machine)...